File: src/games/my-sample-rpg/playerEquipment.ts
export type PlayerEquipmentSlotId =
  | 'weapon'
  | 'armor'
  | 'hat'
  | 'boots'
  | 'accessory'

export type PlayerEquipmentIconKey =
  | 'tiny-dungeon-weapon'
  | 'tiny-knight-gray-helmet'
  | 'tiny-knight-open-helmet'
  | 'town-crate-sword-right'
  | 'weapon-sword'
  | 'weapon-axe'
  | 'weapon-spear'
  | 'weapon-dagger'
  | 'weapon-mace'
  | 'weapon-staff'
  | 'ui-circle-beige'
  | 'ui-check-beige'

export type PlayerEquipmentIcon = {
  key: PlayerEquipmentIconKey
  scale: number
}

export type PlayerEquipmentItem = {
  id: string
  label: string
  level: number
  description: string
}

export type PlayerEquipmentItemDefinition = PlayerEquipmentItem & {
  slotId: PlayerEquipmentSlotId
  icon: PlayerEquipmentIcon
  price: number
}

export type PlayerEquipmentSlot = {
  id: PlayerEquipmentSlotId
  label: string
  item: PlayerEquipmentItem | undefined
}

export type PlayerEquipment = {
  setName: string
  level: number
  slots: PlayerEquipmentSlot[]
}

export const EQUIPMENT_SLOT_LABEL_BY_ID: Record<PlayerEquipmentSlotId, string> = {
  weapon: '무기',
  armor: '옷',
  hat: '모자',
  boots: '신발',
  accessory: '장신구'
}

export const EQUIPMENT_SLOT_IDS: PlayerEquipmentSlotId[] = [
  'weapon',
  'armor',
  'hat',
  'boots',
  'accessory'
]

export const PLAYER_EQUIPMENT_ITEM_DEFINITIONS: PlayerEquipmentItemDefinition[] = [
  {
    id: 'basic-sword',
    slotId: 'weapon',
    label: '기본 무기',
    level: 1,
    description: '초보용 근접 무기',
    price: 120,
    icon: {
      key: 'tiny-dungeon-weapon',
      scale: 1.6
    }
  },
  {
    id: 'basic-armor',
    slotId: 'armor',
    label: '기본 옷',
    level: 1,
    description: '초보용 옷',
    price: 100,
    icon: {
      key: 'tiny-knight-gray-helmet',
      scale: 1.4
    }
  },
  {
    id: 'basic-boots',
    slotId: 'boots',
    label: '기본 신발',
    level: 1,
    description: '초보용 신발',
    price: 80,
    icon: {
      key: 'ui-circle-beige',
      scale: 1.2
    }
  },
  {
    id: 'basic-charm',
    slotId: 'accessory',
    label: '기본 장신구',
    level: 1,
    description: '초보용 장신구',
    price: 60,
    icon: {
      key: 'ui-check-beige',
      scale: 1.2
    }
  },
  {
    id: 'bronze-sword',
    slotId: 'weapon',
    label: '청동 검',
    level: 2,
    description: '대장장이가 만든 단단한 검',
    price: 320,
    icon: {
      key: 'town-crate-sword-right',
      scale: 1.5
    }
  },
  {
    id: 'iron-sword',
    slotId: 'weapon',
    label: '강철 검',
    level: 2,
    description: '균형이 좋은 근접 무기',
    price: 240,
    icon: {
      key: 'weapon-sword',
      scale: 0.08
    }
  },
  {
    id: 'battle-axe',
    slotId: 'weapon',
    label: '전투 도끼',
    level: 2,
    description: '무거운 한손 도끼',
    price: 260,
    icon: {
      key: 'weapon-axe',
      scale: 0.08
    }
  },
  {
    id: 'long-spear',
    slotId: 'weapon',
    label: '장창',
    level: 2,
    description: '거리감을 유지하기 좋은 창',
    price: 280,
    icon: {
      key: 'weapon-spear',
      scale: 0.08
    }
  },
  {
    id: 'quick-dagger',
    slotId: 'weapon',
    label: '단검',
    level: 1,
    description: '빠른 연속 공격용 무기',
    price: 180,
    icon: {
      key: 'weapon-dagger',
      scale: 0.08
    }
  },
  {
    id: 'spiked-mace',
    slotId: 'weapon',
    label: '철퇴',
    level: 3,
    description: '강한 타격을 주는 둔기',
    price: 340,
    icon: {
      key: 'weapon-mace',
      scale: 0.08
    }
  },
  {
    id: 'magic-staff',
    slotId: 'weapon',
    label: '마법 지팡이',
    level: 3,
    description: '마력을 머금은 지팡이',
    price: 360,
    icon: {
      key: 'weapon-staff',
      scale: 0.08
    }
  },
  {
    id: 'iron-armor',
    slotId: 'armor',
    label: '철 옷',
    level: 2,
    description: '초보용을 넘어선 철제 옷',
    price: 260,
    icon: {
      key: 'tiny-knight-open-helmet',
      scale: 1.4
    }
  },
  {
    id: 'Leather_Armor',
    slotId: 'armor',
    label: '가죽 갑옷',
    level: 2,
    description: '몬스터에게서 얻은 가죽 갑옷',
    price: 220,
    icon: {
      key: 'tiny-knight-gray-helmet',
      scale: 1.4
    }
  },
  {
    id: 'Leather_Helmet',
    slotId: 'hat',
    label: '가죽 투구',
    level: 2,
    description: '몬스터에게서 얻은 가죽 투구',
    price: 180,
    icon: {
      key: 'tiny-knight-gray-helmet',
      scale: 1.4
    }
  },
  {
    id: 'Chain_Armor',
    slotId: 'armor',
    label: '사슬 갑옷',
    level: 2,
    description: '몬스터에게서 얻은 사슬 갑옷',
    price: 300,
    icon: {
      key: 'tiny-knight-open-helmet',
      scale: 1.4
    }
  },
  {
    id: 'Chain_Helmet',
    slotId: 'hat',
    label: '사슬 투구',
    level: 2,
    description: '몬스터에게서 얻은 사슬 투구',
    price: 240,
    icon: {
      key: 'tiny-knight-open-helmet',
      scale: 1.4
    }
  },
  {
    id: 'Iron_Armor',
    slotId: 'armor',
    label: '철 갑옷',
    level: 3,
    description: '몬스터에게서 얻은 철 갑옷',
    price: 360,
    icon: {
      key: 'tiny-knight-open-helmet',
      scale: 1.4
    }
  },
  {
    id: 'Iron_Helmet',
    slotId: 'hat',
    label: '철 투구',
    level: 3,
    description: '몬스터에게서 얻은 철 투구',
    price: 300,
    icon: {
      key: 'tiny-knight-open-helmet',
      scale: 1.4
    }
  },
  {
    id: 'leather-boots',
    slotId: 'boots',
    label: '가죽 신발',
    level: 2,
    description: '가볍고 단단한 가죽 신발',
    price: 180,
    icon: {
      key: 'ui-circle-beige',
      scale: 1.2
    }
  },
  {
    id: 'smith-charm',
    slotId: 'accessory',
    label: '수호 부적',
    level: 2,
    description: '대장장이가 준 작은 보호 부적',
    price: 140,
    icon: {
      key: 'ui-check-beige',
      scale: 1.2
    }
  }
]

const PLAYER_EQUIPMENT_ITEM_DEFINITION_BY_ID = new Map(
  PLAYER_EQUIPMENT_ITEM_DEFINITIONS.map((definition) => [
    definition.id,
    definition
  ])
)
const PLAYER_EQUIPMENT_ITEM_DEFINITION_BY_SLOT_ID = new Map<
  PlayerEquipmentSlotId,
  PlayerEquipmentItemDefinition
>()

for (const definition of PLAYER_EQUIPMENT_ITEM_DEFINITIONS) {
  if (!PLAYER_EQUIPMENT_ITEM_DEFINITION_BY_SLOT_ID.has(definition.slotId)) {
    PLAYER_EQUIPMENT_ITEM_DEFINITION_BY_SLOT_ID.set(definition.slotId, definition)
  }
}

const STARTER_WEAPON_ITEM_DEFINITION = PLAYER_EQUIPMENT_ITEM_DEFINITION_BY_ID.get(
  'basic-sword'
)

if (!STARTER_WEAPON_ITEM_DEFINITION) {
  throw new Error('Missing starter weapon definition')
}

export const PLAYER_EQUIPMENT_STARTER_WEAPON_ITEM_DEFINITION: PlayerEquipmentItemDefinition =
  STARTER_WEAPON_ITEM_DEFINITION

export const createInitialPlayerEquipment = (): PlayerEquipment => ({
  setName: '기본 장비',
  level: 1,
  slots: EQUIPMENT_SLOT_IDS.map((slotId) => {
    return {
      id: slotId,
      label: EQUIPMENT_SLOT_LABEL_BY_ID[slotId],
      item: slotId === 'weapon'
        ? createPlayerEquipmentItemFromDefinition(STARTER_WEAPON_ITEM_DEFINITION)
        : undefined
    }
  })
})

export const getPlayerEquipmentSlotLabelById = (
  slotId: PlayerEquipmentSlotId
): string => EQUIPMENT_SLOT_LABEL_BY_ID[slotId]

export const createPlayerEquipmentItemFromDefinition = (
  definition: PlayerEquipmentItemDefinition
): PlayerEquipmentItem => ({
  id: definition.id,
  label: definition.label,
  level: definition.level,
  description: definition.description
})

export const getPlayerEquipmentItemDefinitionById = (
  itemId: string
): PlayerEquipmentItemDefinition | undefined =>
  PLAYER_EQUIPMENT_ITEM_DEFINITION_BY_ID.get(itemId)

export const getPlayerEquipmentItemDefinitionBySlotId = (
  slotId: PlayerEquipmentSlotId
): PlayerEquipmentItemDefinition | undefined =>
  PLAYER_EQUIPMENT_ITEM_DEFINITION_BY_SLOT_ID.get(slotId)

export const getPlayerEquipmentSlotById = (
  equipment: PlayerEquipment,
  slotId: PlayerEquipmentSlotId
): PlayerEquipmentSlot | undefined =>
  equipment.slots.find((slot) => slot.id === slotId)

export const findPlayerEquipmentSlotIndexById = (
  equipment: PlayerEquipment,
  slotId: PlayerEquipmentSlotId
): number => equipment.slots.findIndex((slot) => slot.id === slotId)

export const setPlayerEquipmentSlot = ({
  equipment,
  slotId,
  item
}: {
  equipment: PlayerEquipment
  slotId: PlayerEquipmentSlotId
  item: PlayerEquipmentItem
}): PlayerEquipment => {
  const slotIndex = findPlayerEquipmentSlotIndexById(equipment, slotId)

  if (slotIndex < 0) {
    throw new Error(`Invalid equipment slot id ${slotId}`)
  }

  const slots = [...equipment.slots]
  const slot = slots[slotIndex]

  slots[slotIndex] = {
    ...slot,
    item
  }

  return {
    ...equipment,
    slots
  }
}

export const clearPlayerEquipmentSlot = ({
  equipment,
  slotId
}: {
  equipment: PlayerEquipment
  slotId: PlayerEquipmentSlotId
}): PlayerEquipment => {
  const slotIndex = findPlayerEquipmentSlotIndexById(equipment, slotId)

  if (slotIndex < 0) {
    throw new Error(`Invalid equipment slot id ${slotId}`)
  }

  const slots = [...equipment.slots]
  const slot = slots[slotIndex]

  slots[slotIndex] = {
    ...slot,
    item: undefined
  }

  return {
    ...equipment,
    slots
  }
}

File: src/games/my-sample-rpg/assets/lua/player-equipment.lua
-- TS playerEquipment.ts → Lua. 장비 슬롯(weapon/armor/hat/boots/accessory) 객체를 다룬다(JSON 마샬링).
-- 빈 슬롯의 item 은 TS에서 undefined → JSON null, Lua에선 json_null 센티넬로 보존하고 래퍼가 undefined로 정규화한다.
-- slot_index 는 JS 0-based, Lua slots 배열은 1-based. findIndex 미발견은 TS와 동일하게 -1을 반환한다.
--
-- 데이터 소유는 TS(단일 출처): 슬롯 라벨 맵, 슬롯 id 순서, 스타터 무기 정의, 아이템 정의 배열은
-- 모두 호출 시 인자로 전달된다(여기엔 큰 데이터 테이블을 복제하지 않는다).

-- profile/slot 처럼 문자열 키 테이블을 얕게 복사한다(TS의 spread 패턴).
local function shallow_copy(t)
  local out = {}
  for k, v in pairs(t) do
    out[k] = v
  end
  return out
end

-- slots 시퀀스를 새 배열로 복사한다(요소는 그대로 — TS의 [...slots]).
local function copy_slots(slots)
  local out = json_array({})
  for i = 1, #slots do
    out[i] = slots[i]
  end
  return out
end

-- ── createPlayerEquipmentItemFromDefinition ──
-- definition 에서 PlayerEquipmentItem 으로 추리는 순수 변환.
function equipment_create_item_from_definition(definition)
  return {
    id = definition.id,
    label = definition.label,
    level = definition.level,
    description = definition.description
  }
end

-- ── createInitialPlayerEquipment ──
-- slot_ids: 슬롯 id 순서(배열), slot_label_by_id: id→라벨 맵(객체), starter_weapon: 스타터 무기 정의.
-- weapon 슬롯만 스타터 무기 아이템, 나머지는 빈 슬롯(json_null → 래퍼에서 undefined).
function equipment_create_initial(slot_ids, slot_label_by_id, starter_weapon)
  local slots = json_array({})
  for i = 1, #slot_ids do
    local slot_id = slot_ids[i]
    local item
    if slot_id == 'weapon' then
      item = equipment_create_item_from_definition(starter_weapon)
    else
      item = json_null
    end
    slots[i] = {
      id = slot_id,
      label = slot_label_by_id[slot_id],
      item = item
    }
  end

  return { setName = '기본 장비', level = 1, slots = slots }
end

-- ── getPlayerEquipmentSlotLabelById ──
function equipment_slot_label_by_id(slot_label_by_id, slot_id)
  return slot_label_by_id[slot_id]
end

-- ── getPlayerEquipmentItemDefinitionById ── (정의 배열에서 id 일치 1건, 미발견 시 json_null)
function equipment_item_definition_by_id(definitions, item_id)
  for i = 1, #definitions do
    if definitions[i].id == item_id then
      return definitions[i]
    end
  end
  return json_null
end

-- ── getPlayerEquipmentItemDefinitionBySlotId ──
-- TS는 slotId 별 첫 등장 정의를 맵에 담는다 → 배열 순서상 slotId 가 처음 나오는 정의를 반환.
function equipment_item_definition_by_slot_id(definitions, slot_id)
  for i = 1, #definitions do
    if definitions[i].slotId == slot_id then
      return definitions[i]
    end
  end
  return json_null
end

-- ── findPlayerEquipmentSlotIndexById ── (TS findIndex: 미발견 -1, 발견 시 0-based)
function equipment_find_slot_index_by_id(equipment, slot_id)
  for i = 1, #equipment.slots do
    if equipment.slots[i].id == slot_id then
      return i - 1
    end
  end
  return -1
end

-- ── getPlayerEquipmentSlotById ── (미발견 시 json_null)
function equipment_slot_by_id(equipment, slot_id)
  for i = 1, #equipment.slots do
    if equipment.slots[i].id == slot_id then
      return equipment.slots[i]
    end
  end
  return json_null
end

-- ── setPlayerEquipmentSlot ── 대상 슬롯의 item 을 새 item 으로 교체(슬롯 다른 키 보존).
function equipment_set_slot(equipment, slot_id, item)
  local slot_index = equipment_find_slot_index_by_id(equipment, slot_id)

  if slot_index < 0 then
    error('Invalid equipment slot id ' .. slot_id)
  end

  local slots = copy_slots(equipment.slots)
  local next_slot = shallow_copy(slots[slot_index + 1])
  next_slot.item = item
  slots[slot_index + 1] = next_slot

  local out = shallow_copy(equipment)
  out.slots = slots
  return out
end

-- ── clearPlayerEquipmentSlot ── 대상 슬롯의 item 을 비운다(json_null → 래퍼에서 undefined).
function equipment_clear_slot(equipment, slot_id)
  local slot_index = equipment_find_slot_index_by_id(equipment, slot_id)

  if slot_index < 0 then
    error('Invalid equipment slot id ' .. slot_id)
  end

  local slots = copy_slots(equipment.slots)
  local next_slot = shallow_copy(slots[slot_index + 1])
  next_slot.item = json_null
  slots[slot_index + 1] = next_slot

  local out = shallow_copy(equipment)
  out.slots = slots
  return out
end
